File: services/identity-service/src/test/java/br/com/deltaglobalbank/identity/features/manageModules/ManageModuleTests.java
package br.com.deltaglobalbank.identity.features.manageModules;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

import br.com.deltaglobalbank.identity.domain.module.Module;
import br.com.deltaglobalbank.identity.domain.module.ModuleCode;
import br.com.deltaglobalbank.identity.domain.module.ModuleNotFoundException;
import br.com.deltaglobalbank.identity.domain.module.ModuleRepository;
import br.com.deltaglobalbank.identity.domain.module.TenantModule;
import br.com.deltaglobalbank.identity.domain.module.TenantModuleRepository;
import br.com.deltaglobalbank.identity.domain.tenant.Tenant;
import br.com.deltaglobalbank.identity.domain.tenant.TenantRepository;
import br.com.deltaglobalbank.identity.domain.tenant.TenantSlug;
import br.com.deltaglobalbank.identity.domain.user.TenantNotFoundException;
import br.com.deltaglobalbank.identity.features.modules.disableModule.DisableTenantModuleCommand;
import br.com.deltaglobalbank.identity.features.modules.disableModule.DisableTenantModuleUseCase;
import br.com.deltaglobalbank.identity.features.modules.enableModule.EnableModuleResponse;
import br.com.deltaglobalbank.identity.features.modules.enableModule.EnableTenantModuleCommand;
import br.com.deltaglobalbank.identity.features.modules.enableModule.EnableTenantModuleUseCase;
import br.com.deltaglobalbank.identity.features.modules.listModules.ListTenantModulesResponse;
import br.com.deltaglobalbank.identity.features.modules.listModules.ListTenantModulesUseCase;
import br.com.deltaglobalbank.identity.features.modules.listModules.TenantModuleListItem;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;

class ManageModuleTests {

    private final TenantRepository tenantRepository = mock(TenantRepository.class);
    private final ModuleRepository moduleRepository = mock(ModuleRepository.class);
    private final TenantModuleRepository tenantModuleRepository = mock(TenantModuleRepository.class);

    private final EnableTenantModuleUseCase enableUseCase =
        new EnableTenantModuleUseCase(tenantRepository, moduleRepository, tenantModuleRepository);
    private final DisableTenantModuleUseCase disableUseCase =
        new DisableTenantModuleUseCase(tenantRepository, moduleRepository, tenantModuleRepository);
    private final ListTenantModulesUseCase listUseCase =
        new ListTenantModulesUseCase(tenantRepository, moduleRepository, tenantModuleRepository);

    private final UUID tenantId = UUID.randomUUID();
    private final UUID moduleId = UUID.randomUUID();
    private final UUID cardModuleId = UUID.randomUUID();

    private final Tenant tenant = Tenant.create(tenantId, "Empresa Teste", new TenantSlug("empresa-teste"));

    private final Module lendingModule =
        new Module(moduleId, new ModuleCode("lending"), "Empréstimos", "teste", Instant.now());

    private final Module cardModule =
        new Module(cardModuleId, new ModuleCode("card"), "Cartão", "teste", Instant.now());

    @Test
    void moduleShouldBeEnabled() {
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);
        when(moduleRepository.findByCode(new ModuleCode("lending"))).thenReturn(lendingModule);
        when(tenantModuleRepository.findByTenantIdAndModuleId(tenantId, moduleId)).thenReturn(null);

        ArgumentCaptor<TenantModule> savedCaptor = ArgumentCaptor.forClass(TenantModule.class);
        when(tenantModuleRepository.save(savedCaptor.capture())).thenAnswer(inv -> inv.getArgument(0));

        EnableModuleResponse response = enableUseCase.execute(new EnableTenantModuleCommand(tenantId, "lending"));

        assertEquals("lending", response.moduleCode());
        assertTrue(response.enabled());
        assertNotNull(response.enabledAt());
        assertTrue(savedCaptor.getValue().isEnabled());
    }

    @Test
    void moduleShouldBeReactivated() {
        TenantModule existingTenantModule = TenantModule.create(UUID.randomUUID(), tenantId, cardModuleId);
        existingTenantModule.disable();
        assertFalse(existingTenantModule.isEnabled());

        when(tenantRepository.findById(tenantId)).thenReturn(tenant);
        when(moduleRepository.findByCode(new ModuleCode("card"))).thenReturn(cardModule);
        when(tenantModuleRepository.findByTenantIdAndModuleId(tenantId, cardModuleId)).thenReturn(existingTenantModule);
        when(tenantModuleRepository.save(any())).thenAnswer(inv -> inv.getArgument(0));

        EnableModuleResponse response = enableUseCase.execute(new EnableTenantModuleCommand(tenantId, "card"));

        assertTrue(response.enabled());
        assertTrue(existingTenantModule.isEnabled());
    }

    @Test
    void shouldReturnModuleNotFoundWhenModuleDoesNotExist() {
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);
        when(moduleRepository.findByCode(new ModuleCode("inexistente"))).thenReturn(null);

        assertThrows(ModuleNotFoundException.class,
            () -> enableUseCase.execute(new EnableTenantModuleCommand(tenantId, "inexistente")));
    }

    @Test
    void shouldReturnTenantNotFoundWhenTenantDoesNotExist() {
        UUID unknownTenantId = UUID.randomUUID();
        when(tenantRepository.findById(unknownTenantId)).thenReturn(null);

        assertThrows(TenantNotFoundException.class,
            () -> enableUseCase.execute(new EnableTenantModuleCommand(unknownTenantId, "lending")));
    }

    @Test
    void moduleShouldBeDisabled() {
        TenantModule existingTenantModule = TenantModule.create(UUID.randomUUID(), tenantId, moduleId);
        assertTrue(existingTenantModule.isEnabled());

        when(tenantRepository.findById(tenantId)).thenReturn(tenant);
        when(moduleRepository.findByCode(new ModuleCode("lending"))).thenReturn(lendingModule);
        when(tenantModuleRepository.findByTenantIdAndModuleId(tenantId, moduleId)).thenReturn(existingTenantModule);
        when(tenantModuleRepository.save(any())).thenAnswer(inv -> inv.getArgument(0));

        disableUseCase.execute(new DisableTenantModuleCommand(tenantId, "lending"));

        assertFalse(existingTenantModule.isEnabled());
        verify(tenantModuleRepository, times(1)).save(existingTenantModule);
    }

    @Test
    void disablingModuleWithoutEntryIsIdempotent() {
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);
        when(moduleRepository.findByCode(new ModuleCode("card"))).thenReturn(cardModule);
        when(tenantModuleRepository.findByTenantIdAndModuleId(tenantId, cardModuleId)).thenReturn(null);

        disableUseCase.execute(new DisableTenantModuleCommand(tenantId, "card"));

        verify(tenantModuleRepository, times(0)).save(any());
    }

    @Test
    void moduleShouldBeListed() {
        TenantModule lendingTm = TenantModule.create(UUID.randomUUID(), tenantId, moduleId);
        TenantModule cardTm = TenantModule.create(UUID.randomUUID(), tenantId, cardModuleId);
        cardTm.disable();

        when(tenantRepository.findById(tenantId)).thenReturn(tenant);
        when(moduleRepository.findAll()).thenReturn(List.of(lendingModule, cardModule));
        when(tenantModuleRepository.findAllByTenantId(tenantId)).thenReturn(List.of(lendingTm, cardTm));

        ListTenantModulesResponse response = listUseCase.execute(tenantId);

        assertEquals(2, response.items().size());
        TenantModuleListItem lendingItem = response.items().stream()
            .filter(it -> it.moduleCode().equals("lending")).findFirst().orElseThrow();
        TenantModuleListItem cardItem = response.items().stream()
            .filter(it -> it.moduleCode().equals("card")).findFirst().orElseThrow();
        assertTrue(lendingItem.enabled());
        assertFalse(cardItem.enabled());
    }

    @Test
    void listShouldShowModuleAsDisabledWhenTenantHasNoEntry() {
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);
        when(moduleRepository.findAll()).thenReturn(List.of(lendingModule));
        when(tenantModuleRepository.findAllByTenantId(tenantId)).thenReturn(List.of());

        ListTenantModulesResponse response = listUseCase.execute(tenantId);

        assertEquals(1, response.items().size());
        assertFalse(response.items().get(0).enabled());
    }
}

File: services/identity-service/src/test/kotlin/br/com/deltaglobalbank/identity/features/manageModules/ManageModuleTests.kt
package br.com.deltaglobalbank.identity.features.manageModules

import br.com.deltaglobalbank.identity.domain.module.Module
import br.com.deltaglobalbank.identity.domain.module.ModuleCode
import br.com.deltaglobalbank.identity.domain.module.ModuleNotFoundException
import br.com.deltaglobalbank.identity.domain.module.ModuleRepository
import br.com.deltaglobalbank.identity.domain.module.TenantModule
import br.com.deltaglobalbank.identity.domain.module.TenantModuleRepository
import br.com.deltaglobalbank.identity.domain.tenant.Tenant
import br.com.deltaglobalbank.identity.domain.tenant.TenantRepository
import br.com.deltaglobalbank.identity.domain.tenant.TenantSlug
import br.com.deltaglobalbank.identity.domain.user.TenantNotFoundException
import br.com.deltaglobalbank.identity.features.modules.disableModule.DisableTenantModuleCommand
import br.com.deltaglobalbank.identity.features.modules.disableModule.DisableTenantModuleUseCase
import br.com.deltaglobalbank.identity.features.modules.enableModule.EnableTenantModuleCommand
import br.com.deltaglobalbank.identity.features.modules.enableModule.EnableTenantModuleUseCase
import br.com.deltaglobalbank.identity.features.modules.listModules.ListTenantModulesUseCase
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertNotNull
import org.junit.jupiter.api.assertThrows
import java.time.Instant
import java.util.UUID
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ManageModuleTests {

    private val tenantRepository = mockk<TenantRepository>()
    private val moduleRepository = mockk<ModuleRepository>()
    private val tenantModuleRepository = mockk<TenantModuleRepository>()

    private val enableUseCase = EnableTenantModuleUseCase(
        tenantRepository, moduleRepository, tenantModuleRepository
    )
    private val disableUseCase = DisableTenantModuleUseCase(
        tenantRepository, moduleRepository, tenantModuleRepository
    )
    private val listUseCase = ListTenantModulesUseCase(
        tenantRepository, moduleRepository, tenantModuleRepository
    )

    private val tenantId = UUID.randomUUID()
    private val moduleId = UUID.randomUUID()
    private val cardModuleId = UUID.randomUUID()

    private val tenant = Tenant.create(
        id = tenantId,
        name = "Empresa Teste",
        slug = TenantSlug("empresa-teste")
    )

    private val lendingModule = Module(
        id = moduleId,
        code = ModuleCode("lending"),
        name = "Empréstimos",
        description = "teste",
        createdAt = Instant.now()
    )

    private val cardModule = Module(
        id = cardModuleId,
        code = ModuleCode("card"),
        name = "Cartão",
        description = "teste",
        createdAt = Instant.now()
    )

    @Test
    fun `module should be enabled`() {
        every { tenantRepository.findById(tenantId) } returns tenant
        every { moduleRepository.findByCode(ModuleCode("lending")) } returns lendingModule
        every { tenantModuleRepository.findByTenantIdAndModuleId(tenantId, moduleId) } returns null

        val savedSlot = slot<TenantModule>()
        every { tenantModuleRepository.save(capture(savedSlot)) } answers { savedSlot.captured }

        val response = enableUseCase.execute(EnableTenantModuleCommand(tenantId, "lending"))

        assertEquals("lending", response.moduleCode)
        assertTrue(response.enabled)
        assertNotNull(response.enabledAt)
        assertTrue(savedSlot.captured.isEnabled())
    }

    @Test
    fun `module should be reactivated`() {
        val existingTenantModule = TenantModule.create(
            id = UUID.randomUUID(),
            tenantId = tenantId,
            moduleId = cardModuleId
        )
        existingTenantModule.disable()
        assertFalse(existingTenantModule.isEnabled())

        every { tenantRepository.findById(tenantId) } returns tenant
        every { moduleRepository.findByCode(ModuleCode("card")) } returns cardModule
        every {
            tenantModuleRepository.findByTenantIdAndModuleId(
                tenantId,
                cardModuleId
            )
        } returns existingTenantModule
        every { tenantModuleRepository.save(any()) } answers { firstArg() }

        val response = enableUseCase.execute(
            EnableTenantModuleCommand(tenantId = tenantId, moduleCode = "card")
        )

        assertTrue(response.enabled)
        assertTrue(existingTenantModule.isEnabled())
    }

    @Test
    fun `should return module_not_found when module does not exist`() {
        every { tenantRepository.findById(tenantId) } returns tenant
        every { moduleRepository.findByCode(ModuleCode("inexistente")) } returns null

        assertThrows<ModuleNotFoundException> {
            enableUseCase.execute(
                EnableTenantModuleCommand(tenantId = tenantId, moduleCode = "inexistente")
            )
        }
    }

    @Test
    fun `should return tenant_not_found when tenant does not exist`() {
        val unknownTenantId = UUID.randomUUID()
        every { tenantRepository.findById(unknownTenantId) } returns null

        assertThrows<TenantNotFoundException> {
            enableUseCase.execute(
                EnableTenantModuleCommand(tenantId = unknownTenantId, moduleCode = "lending")
            )
        }
    }

    @Test
    fun `module should be disabled`() {
        val existingTenantModule = TenantModule.create(
            id = UUID.randomUUID(),
            tenantId = tenantId,
            moduleId = moduleId
        )
        assertTrue(existingTenantModule.isEnabled())

        every { tenantRepository.findById(tenantId) } returns tenant
        every { moduleRepository.findByCode(ModuleCode("lending")) } returns lendingModule
        every {
            tenantModuleRepository.findByTenantIdAndModuleId(
                tenantId,
                moduleId
            )
        } returns existingTenantModule
        every { tenantModuleRepository.save(any()) } answers { firstArg() }

        disableUseCase.execute(
            DisableTenantModuleCommand(tenantId = tenantId, moduleCode = "lending")
        )

        assertFalse(existingTenantModule.isEnabled())
        verify { tenantModuleRepository.save(existingTenantModule) }
    }

    @Test
    fun `disabling module without entry is idempotent`() {
        every { tenantRepository.findById(tenantId) } returns tenant
        every { moduleRepository.findByCode(ModuleCode("card")) } returns cardModule
        every {
            tenantModuleRepository.findByTenantIdAndModuleId(
                tenantId,
                cardModuleId
            )
        } returns null

        disableUseCase.execute(
            DisableTenantModuleCommand(tenantId = tenantId, moduleCode = "card")
        )

        verify(exactly = 0) { tenantModuleRepository.save(any()) }
    }

    @Test
    fun `module should be listed`() {
        val lendingTm = TenantModule.create(
            id = UUID.randomUUID(),
            tenantId = tenantId,
            moduleId = moduleId
        )
        val cardTm = TenantModule.create(
            id = UUID.randomUUID(),
            tenantId = tenantId,
            moduleId = cardModuleId
        )
        cardTm.disable()

        every { tenantRepository.findById(tenantId) } returns tenant
        every { moduleRepository.findAll() } returns listOf(lendingModule, cardModule)
        every { tenantModuleRepository.findAllByTenantId(tenantId) } returns listOf(
            lendingTm,
            cardTm
        )

        val response = listUseCase.execute(tenantId)

        assertEquals(2, response.items.size)
        val lendingItem = response.items.first { it.moduleCode == "lending" }
        val cardItem = response.items.first { it.moduleCode == "card" }
        assertTrue(lendingItem.enabled)
        assertFalse(cardItem.enabled)
    }

    @Test
    fun `list should show module as disabled when tenant has no entry`() {
        every { tenantRepository.findById(tenantId) } returns tenant
        every { moduleRepository.findAll() } returns listOf(lendingModule)
        every { tenantModuleRepository.findAllByTenantId(tenantId) } returns emptyList()

        val response = listUseCase.execute(tenantId)

        assertEquals(1, response.items.size)
        assertFalse(response.items.first().enabled)
    }
}